src/stream/src/executor/backfill.rs

```diff
@@ -111,57 +121,114 @@ where
         let is_snapshot_empty: bool = {
-            let snapshot = Self::snapshot_read(&self.table, init_epoch, None, false);
+            let snapshot = Self::snapshot_read(&self.upstream_table, init_epoch, None, false);
             pin_mut!(snapshot);
             snapshot.try_next().await?.unwrap().is_none()
         };
-        let to_backfill = to_create_mv && !is_snapshot_empty;
 
-        if to_create_mv && is_snapshot_empty {
-            // Directly finish the progress as the snapshot is empty.
-            self.progress.finish(first_barrier.epoch.curr);
-        }
+        // Backfill not Finished: Need to Backfill if snapshot is not empty.
+        // Backfill is  Finished: No Need to Backfill.
+        let to_backfill = if !is_finished {
+            !is_snapshot_empty
+        } else {
+            false
+        };
+
+        // Current position of the upstream_table storage primary key.
+        // `None` means it starts from the beginning.
+        let mut current_pos: Option<OwnedRow> = None;
+
+        // Use these to persist state.
+        // They contain the backfill position,
+        // as well as the progress.
+        // However, they do not contain the vnode key at index 0.
+        // That is filled in when we flush the state table.
+        let mut current_state: Vec<Datum> = vec![None; state_len];
+        let mut old_state: Option<Vec<Datum>> = None;
 
         // The first barrier message should be propagated.
         yield Message::Barrier(first_barrier);
 
+        // If no need backfill, but state was still "unfinished" we need to finish it.
+        // So we just update the state + progress to meta at the next barrier to finish progress,
+        // and forward other messages.
+        //
+        // Reason for persisting on second barrier rather than first:
+        // We can't update meta with progress as finished until state_table
+        // has been updated.
+        // We also can't update state_table in first epoch, since state_table
+        // expects to have been initialized in previous epoch.
+        if !is_finished && !to_backfill {
+            while let Some(Ok(msg)) = upstream.next().await {
+                if let Some(msg) = Self::mapping_message(msg, &self.output_indices) {
+                    match &msg {
+                        Message::Barrier(barrier) => {
+                            current_pos =
+                                Self::construct_initial_finished_state(pk_in_output_indices.len());
+                            Self::persist_state(
+                                barrier.epoch,
+                                &mut self.state_table,
+                                true,
+                                &current_pos,
+                                &mut old_state,
+                                &mut current_state,
+                            )
+                            .await?;
+                            self.progress.finish(barrier.epoch.curr);
```

Comment:
The logic here is to persist finished state to state table on second barrier. Could you please explain what "optimization" are you referring to? Didn't quite get it.

Reply:
> We can support init state table with state? Although I'm not sure what changes that would require.

I guess this might be feasible. For backfill, we
- receive the first barrier
- grab the "prev epoch" from it
- write finished states with this epoch
- `finish`, commit, then yield the barrier

In #7009, we've mentioned that the optimization can't work correctly even if there're only a few rows in the table. This is because we have to maintain the total order of data and barriers, so the downstream stateful executors must also handle data (then write states) before receiving the first barrier, which requires a considerable amount of changes that are not as easy as we described for backfill above.

However, there seems an assertion for no data in the very first epoch after cluster launches. Not sure whether there're more assumptions based on this. cc @hzxa21 

https://github.com/risingwavelabs/risingwave/blob/318714b2a50befdb42cd7ec9e4c6590dd43d9e6d/src/meta/src/barrier/mod.rs#L896-L901

Reply:
https://github.com/risingwavelabs/risingwave/issues/9972